File: ArmaReforgerServerTool/Models/SavedState.cs
/******************************************************************************
 * File Name:    SavedState.cs
 * Project:      Longbow
 * Description:  This file contains the SavedState class which is a model
 *               which represents any parameters that can save state and
 *               should survive exits and restarts
 * 
 * Author:       Bradley Newman
 ******************************************************************************/

using ReforgerServerApp.Utils;
using System.Text.Json;
using System.Text.Json.Serialization;
using static ReforgerServerApp.Utils.JsonUtils;

namespace Longbow.Models
{
  [JsonConverter(typeof(SavedStateConverter))]
  internal class SavedState
  {
    // Default Advanced Server Parameters
    public static readonly AdvancedSetting DEFAULT_MAX_FPS = new("maxFPS", 60, true);
    public static readonly AdvancedSetting DEFAULT_VERIFY_AND_REPAIR_ADDONS = new("addonsRepair", true);
    public static readonly AdvancedSetting DEFAULT_USE_UPNP = new("useUpnp", true);
    public static readonly AdvancedSetting DEFAULT_KEEP_SERVER_UPDATED = new("keepServerUpdated", true);
    public static readonly AdvancedSetting DEFAULT_USE_EXPERIMENTAL = new("useExperimental", false);
    public static readonly AdvancedSetting DEFAULT_LOAD_SESSION_SAVE = new("loadSessionSave", "", false);
    public static readonly AdvancedSetting DEFAULT_AUTO_RESTART_ON_CRASH = new("autoRestartOnCrash", false);
    public static readonly AdvancedSetting DEFAULT_AUTO_RELOAD = new("autoreload", 10, false);
    public static readonly AdvancedSetting DEFAULT_NO_BACKEND = new("noBackend", false);
    public static readonly AdvancedSetting DEFAULT_AUTO_SHUTDOWN = new("autoShutdown", false);
    public static readonly AdvancedSetting DEFAULT_LOG_VOTING = new("logVoting", false);
    public static readonly AdvancedSetting DEFAULT_OVERRIDE_PORT = new("bindPort", 2001, false);
    public static readonly AdvancedSetting DEFAULT_NETWORK_DYNAMIC_SIM = new("nds", 2, false);
    public static readonly AdvancedSetting DEFAULT_SPATIAL_MAP_RES = new("nwkResolution", 500, false);
    public static readonly AdvancedSetting DEFAULT_STAGGERING_BUDGET = new("staggeringBudget", 5000, false);
    public static readonly AdvancedSetting DEFAULT_STREAMING_BUDGET = new("streamingBudget", 500, false);
    public static readonly AdvancedSetting DEFAULT_STREAMS_DELTA = new("streamsDelta", 100, false);
    public static readonly AdvancedSetting DEFAULT_RPL_TIMEOUT_MS = new("rpl-timeout-ms", 10000, false);
    public static readonly AdvancedSetting DEFAULT_AI_PARTIAL_SIM = new("aiPartialSim", false);
    public static readonly AdvancedSetting DEFAULT_CREATE_DB = new("createDB", false);
    public static readonly AdvancedSetting DEFAULT_DEBUGGER_ADDRESS = new("debugger", "127.0.0.1", false);
    public static readonly AdvancedSetting DEFAULT_DEBUGGER_PORT = new("debuggerPort", 1000, false);
    public static readonly AdvancedSetting DEFAULT_DISABLE_SHADERS_BUILD = new("disableShadersBuild", false);
    public static readonly AdvancedSetting DEFAULT_GENERATE_SHADERS = new("generateShaders", false);
    public static readonly AdvancedSetting DEFAULT_RPL_ENCODE_AS_LONG_JOBS = new("rplEncodeAsLongJobs", false);
    public static readonly AdvancedSetting DEFAULT_JOB_SYS_SHORT_WORKER_COUNT = new("jobsysShortWorkerCount", Utilities.GetNumberAvailableThreads(), false);
    public static readonly AdvancedSetting DEFAULT_JOB_SYS_LONG_WORKER_COUNT = new("jobsysLongWorkerCount", Utilities.GetNumberAvailableThreads() / 2, false);
    public static readonly AdvancedSetting DEFAULT_FREEZE_CHECK = new("freezeCheck", 300, false);
    public static readonly AdvancedSetting DEFAULT_FREEZE_CHECK_MODE = new("freezeCheckMode", "minidump", false);
    public static readonly AdvancedSetting DEFAULT_FORCE_DISABLE_NIGHT_GRAIN = new("forceDisableNightGrain", false);

    public Dictionary<string, AdvancedSetting> advancedSettings { get; set; }
    public string serverLocation { get; set; }

    public SavedState(Dictionary<string, AdvancedSetting> advancedSettings, string serverLocation)
    {
      this.advancedSettings = advancedSettings;
      this.serverLocation = serverLocation;
    }

    public static SavedState Default => new(GetDefaultAdvancedSettings(), string.Empty);

    /// <summary>
    /// Display <c>SavedState</c> in readable Json format.
    /// </summary>
    /// <returns>Json string representation of the <c>SavedState</c></returns>
    public string AsJsonString()
    {
      return JsonSerializer.Serialize(this, new JsonSerializerOptions { WriteIndented = true });
    }

    /// <summary>
    /// Create a default Advanced Settings dictionary
    /// </summary>
    /// <returns>Advanced Settings dictionary containing defaults</returns>
    private static Dictionary<string, AdvancedSetting> GetDefaultAdvancedSettings()
    {
      Dictionary<string, AdvancedSetting> advancedSettings = new();
      advancedSettings[DEFAULT_MAX_FPS.Name] = DEFAULT_MAX_FPS;
      advancedSettings[DEFAULT_VERIFY_AND_REPAIR_ADDONS.Name] = DEFAULT_VERIFY_AND_REPAIR_ADDONS;
      advancedSettings[DEFAULT_USE_UPNP.Name] = DEFAULT_USE_UPNP;
      advancedSettings[DEFAULT_USE_EXPERIMENTAL.Name] = DEFAULT_USE_EXPERIMENTAL;
      advancedSettings[DEFAULT_KEEP_SERVER_UPDATED.Name] = DEFAULT_KEEP_SERVER_UPDATED;
      advancedSettings[DEFAULT_MAX_FPS.Name] = DEFAULT_MAX_FPS;
      advancedSettings[DEFAULT_VERIFY_AND_REPAIR_ADDONS.Name] = DEFAULT_VERIFY_AND_REPAIR_ADDONS;
      advancedSettings[DEFAULT_USE_UPNP.Name] = DEFAULT_USE_UPNP;
      advancedSettings[DEFAULT_USE_EXPERIMENTAL.Name] = DEFAULT_USE_EXPERIMENTAL;
      advancedSettings[DEFAULT_LOAD_SESSION_SAVE.Name] = DEFAULT_LOAD_SESSION_SAVE;
      advancedSettings[DEFAULT_AUTO_RESTART_ON_CRASH.Name] = DEFAULT_AUTO_RESTART_ON_CRASH;
      advancedSettings[DEFAULT_AUTO_RELOAD.Name] = DEFAULT_AUTO_RELOAD;
      advancedSettings[DEFAULT_NO_BACKEND.Name] = DEFAULT_NO_BACKEND;
      advancedSettings[DEFAULT_AUTO_SHUTDOWN.Name] = DEFAULT_AUTO_SHUTDOWN;
      advancedSettings[DEFAULT_LOG_VOTING.Name] = DEFAULT_LOG_VOTING;
      advancedSettings[DEFAULT_OVERRIDE_PORT.Name] = DEFAULT_OVERRIDE_PORT;
      advancedSettings[DEFAULT_NETWORK_DYNAMIC_SIM.Name] = DEFAULT_NETWORK_DYNAMIC_SIM;
      advancedSettings[DEFAULT_SPATIAL_MAP_RES.Name] = DEFAULT_SPATIAL_MAP_RES;
      advancedSettings[DEFAULT_STAGGERING_BUDGET.Name] = DEFAULT_STAGGERING_BUDGET;
      advancedSettings[DEFAULT_STREAMING_BUDGET.Name] = DEFAULT_STREAMING_BUDGET;
      advancedSettings[DEFAULT_STREAMS_DELTA.Name] = DEFAULT_STREAMS_DELTA;
      advancedSettings[DEFAULT_RPL_TIMEOUT_MS.Name] = DEFAULT_RPL_TIMEOUT_MS;
      advancedSettings[DEFAULT_AI_PARTIAL_SIM.Name] = DEFAULT_AI_PARTIAL_SIM;
      advancedSettings[DEFAULT_CREATE_DB.Name] = DEFAULT_CREATE_DB;
      advancedSettings[DEFAULT_DEBUGGER_ADDRESS.Name] = DEFAULT_DEBUGGER_ADDRESS;
      advancedSettings[DEFAULT_DEBUGGER_PORT.Name] = DEFAULT_DEBUGGER_PORT;
      advancedSettings[DEFAULT_DISABLE_SHADERS_BUILD.Name] = DEFAULT_DISABLE_SHADERS_BUILD;
      advancedSettings[DEFAULT_GENERATE_SHADERS.Name] = DEFAULT_GENERATE_SHADERS;
      advancedSettings[DEFAULT_RPL_ENCODE_AS_LONG_JOBS.Name] = DEFAULT_RPL_ENCODE_AS_LONG_JOBS;
      advancedSettings[DEFAULT_JOB_SYS_SHORT_WORKER_COUNT.Name] = DEFAULT_JOB_SYS_SHORT_WORKER_COUNT;
      advancedSettings[DEFAULT_JOB_SYS_LONG_WORKER_COUNT.Name] = DEFAULT_JOB_SYS_LONG_WORKER_COUNT;
      advancedSettings[DEFAULT_FREEZE_CHECK.Name] = DEFAULT_FREEZE_CHECK;
      advancedSettings[DEFAULT_FREEZE_CHECK_MODE.Name] = DEFAULT_FREEZE_CHECK_MODE;
      advancedSettings[DEFAULT_FORCE_DISABLE_NIGHT_GRAIN.Name] = DEFAULT_FORCE_DISABLE_NIGHT_GRAIN;
      return advancedSettings;
    }
  }
}

File: app/src-tauri/src/models/saved_state.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use super::advanced_setting::{AdvancedSetting, AdvancedSettingValue};

/// Number of available hardware threads, capped at 16. Mirrors
/// `Utilities.GetNumberAvailableThreads()` from the C# original.
///
/// NOTE: this is intentionally a local, private helper (not `crate::util`) to avoid a
/// cross-module dependency while `util` is being ported in parallel elsewhere.
fn available_threads() -> usize {
    std::cmp::min(
        std::thread::available_parallelism()
            .map(|n| n.get())
            .unwrap_or(4),
        16,
    )
}

/// Represents any parameters that can save state and should survive exits and restarts.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SavedState {
    #[serde(default)]
    pub advanced_settings: HashMap<String, AdvancedSetting>,
    #[serde(default)]
    pub server_location: String,
}

impl SavedState {
    /// Builds the default `SavedState`: the ~30 default advanced settings (mirroring the
    /// C# `SavedState.Default`) and an empty server location.
    pub fn default_state() -> SavedState {
        let mut advanced_settings = HashMap::new();

        let threads = available_threads();

        let defaults = [
            AdvancedSetting::new("maxFPS", AdvancedSettingValue::Int(60), true),
            AdvancedSetting::new_switch("addonsRepair", true),
            AdvancedSetting::new_switch("useUpnp", true),
            AdvancedSetting::new_switch("keepServerUpdated", true),
            AdvancedSetting::new_switch("useExperimental", false),
            AdvancedSetting::new("loadSessionSave", AdvancedSettingValue::Str(String::new()), false),
            AdvancedSetting::new_switch("autoRestartOnCrash", false),
            AdvancedSetting::new("autoreload", AdvancedSettingValue::Int(10), false),
            AdvancedSetting::new_switch("noBackend", false),
            AdvancedSetting::new_switch("autoShutdown", false),
            AdvancedSetting::new_switch("logVoting", false),
            AdvancedSetting::new("bindPort", AdvancedSettingValue::Int(2001), false),
            AdvancedSetting::new("nds", AdvancedSettingValue::Int(2), false),
            AdvancedSetting::new("nwkResolution", AdvancedSettingValue::Int(500), false),
            AdvancedSetting::new("staggeringBudget", AdvancedSettingValue::Int(5000), false),
            AdvancedSetting::new("streamingBudget", AdvancedSettingValue::Int(500), false),
            AdvancedSetting::new("streamsDelta", AdvancedSettingValue::Int(100), false),
            AdvancedSetting::new("rpl-timeout-ms", AdvancedSettingValue::Int(10000), false),
            AdvancedSetting::new_switch("aiPartialSim", false),
            AdvancedSetting::new_switch("createDB", false),
            AdvancedSetting::new(
                "debugger",
                AdvancedSettingValue::Str("127.0.0.1".to_string()),
                false,
            ),
            AdvancedSetting::new("debuggerPort", AdvancedSettingValue::Int(1000), false),
            AdvancedSetting::new_switch("disableShadersBuild", false),
            AdvancedSetting::new_switch("generateShaders", false),
            AdvancedSetting::new_switch("rplEncodeAsLongJobs", false),
            AdvancedSetting::new(
                "jobsysShortWorkerCount",
                AdvancedSettingValue::Int(threads as i64),
                false,
            ),
            AdvancedSetting::new(
                "jobsysLongWorkerCount",
                AdvancedSettingValue::Int((threads / 2) as i64),
                false,
            ),
            AdvancedSetting::new("freezeCheck", AdvancedSettingValue::Int(300), false),
            AdvancedSetting::new(
                "freezeCheckMode",
                AdvancedSettingValue::Str("minidump".to_string()),
                false,
            ),
            AdvancedSetting::new_switch("forceDisableNightGrain", false),
        ];

        for setting in defaults {
            advanced_settings.insert(setting.name.clone(), setting);
        }

        SavedState {
            advanced_settings,
            server_location: String::new(),
        }
    }
}

impl Default for SavedState {
    fn default() -> Self {
        Self::default_state()
    }
}
